group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/PartitionWriter.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.coordinator.group.runtime;
+
+import org.apache.kafka.common.KafkaException;
+import org.apache.kafka.common.TopicPartition;
+
+import java.util.List;
+
+/**
+ * A simple interface to write records to Partitions/Logs. It contains the minimum
+ * required for coordinators.
+ *
+ * @param <T> The record type.
+ */
+public interface PartitionWriter<T> {
+
+    /**
+     * Serializer to translate T to bytes.
+     *
+     * @param <T> The record type.
+     */
+    interface Serializer<T> {
+        /**
+         * Serializes the key of the record.
+         */
+        byte[] serializeKey(T record);
+
+        /**
+         * Serializes the value of the record.
+         */
+        byte[] serializeValue(T record);
+    }
+
+    /**
+     * Listener allowing to listen to high watermark changes. This is meant
+     * to be used in conjunction with {{@link PartitionWriter#append(TopicPartition, List)}}.
```

Comment:
my understanding is this gets registered to the Partition. once the hwm is updated, we trigger Listener#onHighWatermarkUpdated but we don't use in PartitionWriter#append and I don't see it used anywhere else. 

Is there an example of what listener we actually register?

Reply:
you can check out the next pr to see how this is used.

Reply:
which PR are you referring to: https://github.com/apache/kafka/pull/13795 or https://github.com/apache/kafka/pull/13812?

or do you mean 15/N which i don't see

Reply:
https://github.com/apache/kafka/pull/13795 is the one.

Reply:
ah found HighWatermarkListener. thanks!